File: targets/java/helidon-se/src/main/java/rb/helidon/routes/DomainRoutes.java
package rb.helidon.routes;

import io.helidon.webserver.http.HttpRouting;
import rb.domain.Domain;
import rb.domain.Model.ValidatedOrder;
import rb.domain.Model.ValidatedOrderWithId;
import rb.helidon.Reply;
import rb.helidon.Validation;

/** domain: application-shaped handler work and the write methods. */
public final class DomainRoutes {
  private DomainRoutes() {}

  public static void register(HttpRouting.Builder r) {
    r.get("/domain/orders", (req, res) -> res.send(
        Domain.domainFilter(Query.qint(req, "page", 0), Query.qint(req, "size", 25),
                            Query.qstr(req, "status"))));

    r.post("/domain/orders", (req, res) -> {
      ValidatedOrder v = Validation.validated(Reply.body(req), false);
      res.header("location", Domain.createdLocation());
      res.status(201).send(v);
    });

    r.get("/domain/orders/{oid}",
          (req, res) -> res.send(Domain.getOrder(Reply.param(req, "oid"))));

    r.put("/domain/orders/{oid}", (req, res) -> {
      int id = Domain.getOrder(Reply.param(req, "oid")).id();
      ValidatedOrder v = Validation.validated(Reply.body(req), false);
      res.send(new ValidatedOrderWithId(id, v.customerId(), v.status(), v.lines(),
                                        v.totalCents()));
    });

    r.get("/domain/customers/{cid}/summary",
          (req, res) -> res.send(Domain.domainJoin(Reply.param(req, "cid"))));

    r.get("/domain/regions/{region}/report",
          (req, res) -> res.send(Domain.domainAggregate(Reply.param(req, "region"))));

    r.patch("/domain/customers/{cid}",
            (req, res) -> res.send(Domain.patchCustomer(Reply.param(req, "cid"),
                                                        Reply.body(req))));

    r.delete("/domain/orders/{oid}/lines/{lid}", (req, res) -> {
      Domain.getOrderLine(Reply.param(req, "oid"), Reply.param(req, "lid"));
      Reply.noBody(res, 204);
    });
  }
}

File: targets/dotnet/carter/Routes/DomainRoutes.cs
using System.Text.Json;
using Carter;
using RequestBench.Domain;

namespace RequestBench.CarterTarget.Routes;

/// <summary>domain: application-shaped handler work and the write methods.</summary>
public sealed class DomainRoutes : ICarterModule
{
    public void AddRoutes(IEndpointRouteBuilder app)
    {
        app.MapGet("/domain/orders", (int page, int size, string status, DomainModel d) =>
            d.DomainFilter(page, size, status));

        // Carter's MapPost<T> and MapPut<T> validate the body in Carter's endpoint filter, the
        // same as the body routes, so a refused body never reaches either handler.
        app.MapPost<OrderBody>("/domain/orders", (OrderBody body, DomainModel d,
                                                  HttpResponse response) =>
        {
            ValidatedOrder order = Body.Priced(d, body);
            response.Headers.Location = d.CreatedLocation();
            return Results.Json(order, statusCode: 201);
        });

        app.MapGet("/domain/orders/{oid}", (string oid, DomainModel d) => d.GetOrder(oid));

        app.MapPut<OrderBody>("/domain/orders/{oid}", (string oid, OrderBody body, DomainModel d) =>
        {
            Order existing = d.GetOrder(oid);
            return Results.Ok(Body.Priced(d, body) with { Id = existing.Id });
        });

        app.MapGet("/domain/customers/{cid}/summary", (string cid, DomainModel d) => d.DomainJoin(cid));

        app.MapGet("/domain/regions/{region}/report",
                   (string region, DomainModel d) => d.DomainAggregate(region));

        app.MapPatch("/domain/customers/{cid}",
                     (string cid, JsonElement body, DomainModel d) => d.PatchCustomer(cid, body));

        app.MapDelete("/domain/orders/{oid}/lines/{lid}", (string oid, string lid, DomainModel d) =>
        {
            d.GetOrderLine(oid, lid);
            return Results.NoContent();
        });
    }
}
